File: 0650-2 Keys Keyboard/0650-2 Keys Keyboard.cpp
// DP
class Solution {
public:
    int minSteps(int n) {
        vector<int> dp(1 + n, INT_MAX);
        dp[0] = dp[1] = 0;
        for (int i = 2; i <= n; ++i)
        {
            for (int j = 1; j * 2 <= i; ++j)
            {
                if (i % j == 0)
                {
                    dp[i] = min(dp[i], dp[j] + i / j);
                }                
            }            
        }
        
        return dp[n];
    }
};

// Greedy
class Solution {
public:
    int minSteps(int n) {
        int s = 0;
        for (int i = 2; i <= n; ++i)
        {
            while (n % i == 0)
            {
                s += i;
                n /= i;
            }
        }
        
        return s;
    }
};

File: 0601-0700/0650-2 Keys Keyboard/0650-2 Keys Keyboard.py
class Solution:
    def minSteps(self, n: int) -> int:
        dp = [i for i in range(1 + n)]
        dp[1] = 0
        for i in range(2, 1 + n):
            for j in range(1, i // 2 + 1):
                if i % j == 0:
                    dp[i] = min(dp[i], dp[j] + i // j)
        return dp[-1]

class Solution2:
    def minSteps(self, n: int) -> int:
        step = 0
        i = 2
        while i <= n:
            while n % i == 0:
                step += i
                n //= i
            i += 1
            
        return step
